### ISSUE TITLE
Magento 2.3.0-beta36: Category Products Indexer execution time increased drastically

### ISSUE THREAD

**Issue by jhruehl** [magento/magento2#18647]
### Preconditions (*)
1. Magento 2.3.0-beta36 (Open Source or Commerce) store with around 75k entries in the catalog_product_entity table
2. PHP 7.1

### Steps to reproduce (*)
1. Reindex Category Products Indexer via CLI command

### Expected result (*)
1. Execution time is well below 1 minute (as in 2.3.0-alpha in same local environment)

### Actual result (*)
1. Execution time is around 30 minutes

### Additional note
I ran "show full processlist" during the execution of the indexer and found massive unbatched queries for each store view causing the massive increase of the execution time:

[one_of_the_queries_done_in_catalog_category_product_indexer_since_2-3-0-beta.txt](https://github.com/magento/magento2/files/2483006/one_of_the_queries_done_in_catalog_category_product_indexer_since_2-3-0-beta.txt)

Furthermore the commit where the massive change to the catalog_category_product indexer was done is given here:
https://github.com/magento/magento2/commit/cfefced540cb76da234759e4a135ee2baeaee6b6

I'm guessing the tests for those commits were done purely with sample data having way too few products to actually correctly measure the performance of a change indexer process in bigger stores.

### Comment
This is not the first time I ran into a result of unsatisactory testing, gotten probably through because of unrealistic (too shallow) testing cases. To prevent future mishaps concerning execution times in the future it would be reasonable to have a testing environment with at least 100k product, 50k customers and 30k sales rules or even more. Especially for massive commits like the one I referenced above.

**Comment by magento-engcom-team** [comment]
Hi @jhruehl. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [x] Summary of the issue
- [x] Information on your environment
- [x] Steps to reproduce
- [x] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@jhruehl do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by jhruehl** [comment]
@magento-engcom-team  As stated above this issue cannot be reproduced with a Magento Vanilla Store having just the given optional sample data, because the responsible queries done won't query for even closely as many product ids as present in a true store setup. So I cannot answer this question in any way.

The query given in my uploaded file shows exactly where the cause is. Given the queries done during the reindex process the issue should be most definitely resproducable, if you create around 20 to 50k products in it.

Given the query in my text file the cause is quite obvious: no batching was implemented.

**Comment by jhruehl** [comment]
@magento-engcom-team FYI: The Product EAV Indexer seems to have been screwed up massively, too, somewhere between 2.3.0-alpha and 2.3.0-beta18. It just took 51 minutes instead of around 1 minute now. You definitely need a more appropiate testing environment than the vanilla store with sample data in the future to avoid rendering some core functionalities almost useless by "trusted" commits from the community.

**Comment by magento-engcom-team** [comment]
Hi @TomashKhamlai. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by TomashKhamlai** [comment]
@jhruehl, please update the **Preconditions** or **Steps to reproduce** section, because some bots :smile: can be confused with idea to upgrade from 2.3.0-beta18 in **Preconeditions** to 2.3.0-beta18 in **Steps to reproduce**.

I hope you mean that it is not necessary to update the store, if there is a possibility to have 2 Magento instances with 75k product where one instance is 2.3.0-alpha and another is 2.3.0-beta18. If I am correct, please update the steps because *Update store* step will increase time consumption on manual testing process.

**Comment by jhruehl** [comment]
@TomashKhamlai I just removed the first step in **Steps to reproduce** and updated **Preconditions** to Magento 2.3.0-beta36. Is that enought?

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by shikhamis11** [comment]
I am also facing Same issue at my client site

**Comment by magento-engcom-team** [comment]
@shikhamis11 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-97384`, `MAGETWO-97385` were created

**Comment by magento-engcom-team** [comment]
@jhruehl, thank you for your report.     
    
Unfortunately, we are archiving this ticket now as it did not get much attention from both Magento Community and Core developers for an extended period. This is done in an effort to create a quality, community-driven backlog which will allow us to allocate the required attention more easily.

Please feel free to comment, reopen or create new ticket according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) 
if you are still facing this issue on the latest `2.3-develop` branch. Thank you for collaboration.
